include/ginkgo/core/base/reordered.hpp

```diff
@@ -0,0 +1,244 @@
+
+        /**
+         * The reordering that is to be applied to the system matrix.
+         */
+        std::shared_ptr<const reorder::ReorderingBaseFactory>
+            GKO_FACTORY_PARAMETER_SCALAR(reordering, nullptr);
+
+        /**
+         * The row scaling that is to be applied to the system matrix.
+         */
+        std::shared_ptr<const matrix::Diagonal<value_type>>
+            GKO_FACTORY_PARAMETER_SCALAR(row_scaling, nullptr);
+
+        /**
+         * The column scaling that is to be applied to the system matrix.
+         */
+        std::shared_ptr<const matrix::Diagonal<value_type>>
+            GKO_FACTORY_PARAMETER_SCALAR(col_scaling, nullptr);
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(ScaledReordered, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Creates an empty scaled reordered operator (0x0 operator).
+     */
+    explicit ScaledReordered(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<ScaledReordered>(std::move(exec))
+    {}
+
+    explicit ScaledReordered(const Factory* factory,
+                             std::shared_ptr<const LinOp> system_matrix)
+        : EnableLinOp<ScaledReordered>(factory->get_executor(),
+                                       system_matrix->get_size()),
+          parameters_{factory->get_parameters()}
+    {
+        // For now only support square matrices.
+        GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+
+        auto exec = this->get_executor();
+
+        system_matrix_ = gko::clone(exec, system_matrix);
+
+        // Scale the system matrix if scaling coefficients are provided
+        if (parameters_.row_scaling) {
+            GKO_ASSERT_EQUAL_DIMENSIONS(parameters_.row_scaling,
+                                        system_matrix_);
+            row_scaling_ = parameters_.row_scaling;
+            row_scaling_->apply(system_matrix_.get(), system_matrix_.get());
+        }
+        if (parameters_.col_scaling) {
+            GKO_ASSERT_EQUAL_DIMENSIONS(parameters_.col_scaling,
+                                        system_matrix_);
+            col_scaling_ = parameters_.col_scaling;
+            col_scaling_->rapply(system_matrix_.get(), system_matrix_.get());
+        }
+
+        // If a reordering factory is provided, generate the reordering and
+        // permute the system matrix accordingly.
+        if (parameters_.reordering) {
+            auto reordering = parameters_.reordering->generate(system_matrix_);
+            auto perm = reordering->get_permutation();
+            permutation_ =
+                clone(exec, as<matrix::Permutation<index_type>>(perm));
+            auto permutation_array =
+                make_array_view(exec, permutation_->get_size()[0],
+                                permutation_->get_permutation());
+            system_matrix_ = as<Permutable<index_type>>(system_matrix_)
+                                 ->permute(&permutation_array);
```

Comment:
This requires your reordering to return a `matrix::Permutation`. However, this is not documented and not intuitive (your `ReorderingBaseFactory` needs `get_permutation()` to return a `matrix::Permutation`).
Is there a reason why you don't call the `apply` method on the permutation matrix? That would allow for more formats than just `matrix::Permutation` and would work a lot better with your new interface.

Reply:
I went with an array now, that works for all kinds of permutations and does not assume we have a permutation matrix.